apps/s_client.c

```diff
@@ -2257,6 +2259,62 @@ int s_client_main(int argc, char **argv)
             }
         }
         break;
+    case PROTO_MYSQL:
+        {
+            /*
+             * https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_packets.html
+             * https://dev.mysql.com/doc/internals/en/ssl.html
+             * https://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::HandshakeResponse
+             */
+            static const unsigned char mysql_ssl_request[] = {
+                /* payload_length,   sequence_id */
+                   0x20, 0x00, 0x00, 0x01,
+                /* payload */
+                /* capability flags, CLIENT_SSL always set */
+                   0x85, 0xae, 0x7f, 0x00,
+                /* max-packet size */
+                   0x00, 0x00, 0x00, 0x01,
+                /* character set */
+                   0x21,
+                /* string[23] reserved (all [0]) */
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+            };
+            int bytes;
+            int ssl_capability = 0x800;
+            char *server_version;
+            int server_version_len;
+
+            /*
+             * https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html
+             * Receiving Initial Handshake packet from mysql server
+             */
+            bytes = BIO_read(sbio, sbuf, BUFSIZZ);
+            if (bytes < 1){
+                BIO_printf(bio_err, "BIO_read failed\n");
+                goto shut;
+            } else if (bytes < 4 ||
+                       bytes != (sbuf[0] + sbuf[1]  * 256 + sbuf[2]  * 256 * 256 + 4 ) ) {
```

Comment:
I think we are more accustomed to shifts, so I'd rather suggest `sbuf[0] + (sbuf[1] << 8) + (sbuf[2] << 16)`. BUT! More importantly problem here is that sbuf is pointer to *signed* char (or at least unspecified signedness). This would result in packet length being evaluated as -1 for sequence of \0xff,\0,\0 instead of 255.